### PR TITLE
capture job attempts and associate events with jobs via previous / current event IDs

### PR DESCRIPTION
**Pull request by rgarcia**
https://clever.atlassian.net/browse/INFRA-2575
https://clever.atlassian.net/browse/INFRA-2547

here's an example execution that had a state retry once: https://us-west-2.console.aws.amazon.com/states/home?region=us-west-2#/executions/details/arn:aws:states:us-west-2:589690932525:execution:raf--cil-reliability-dashboard-master--4--start:3956fc73-d95a-43f6-9984-22f5e5f07ffd

before, workflow-manager would return an empty "attempts" array, only reporting the successful job run: 

```
curl https://clever-dev--workflow-manager.int.clever.com/workflows/3956fc73-d95a-43f6-9984-22f5e5f07ffd
{
	"createdAt": "2017-10-09T22:31:02.497Z",
	"id": "3956fc73-d95a-43f6-9984-22f5e5f07ffd",
	"input": "{}",
	"jobs": [
		{
			"attempts": null,
			"createdAt": "2017-10-09T22:31:02.000Z",
			"id": "2",
			"input": "{\"_EXECUTION_NAME\":\"3956fc73-d95a-43f6-9984-22f5e5f07ffd\"}",
			"output": "{\"_EXECUTION_NAME\":\"3956fc73-d95a-43f6-9984-22f5e5f07ffd\",\"done\":true}",
			"startedAt": "2017-10-09T22:31:02.000Z",
			"state": "start",
			"stateResource": {
				"lastUpdated": "2017-10-09T22:31:02.000Z",
				"name": "cil-reliability-dashboard",
				"namespace": "raf",
				"type": "ActivityARN"
			},
			"status": "succeeded",
			"stoppedAt": "2017-10-09T22:33:13.000Z"
		}
	],
	"lastUpdated": "2017-10-09T22:38:00.122Z",
	"namespace": "raf",
	"queue": "development",
	"retries": null,
	"status": "succeeded",
	"workflowDefinition": {
		"createdAt": "2017-10-02T22:20:48.497Z",
		"id": "872ea0df-0a09-4118-868d-bc2324af8078",
		"manager": "step-functions",
		"name": "cil-reliability-dashboard:master",
		"stateMachine": {
			"StartAt": "start",
			"States": {
				"start": {
					"End": true,
					"Resource": "cil-reliability-dashboard",
					"Retry": null,
					"Type": "Task"
				}
			},
			"Version": "1.0"
		},
		"version": 4
	}
}
```

Now it fills in the attempts array:

```
$ curl localhost:8080/workflows/3956fc73-d95a-43f6-9984-22f5e5f07ffd
{
	"createdAt": "2017-10-09T22:31:02.497Z",
	"id": "3956fc73-d95a-43f6-9984-22f5e5f07ffd",
	"input": "{}",
	"jobs": [
		{
			"attempts": [
				{
					"reason": "sfncli.CommandTerminated",
					"startedAt": "2017-10-09T22:31:02.000Z",
					"stoppedAt": "2017-10-09T22:31:20.000Z",
					"taskARN": "ubuntu-xenial"
				}
			],
			"container": "ubuntu-xenial",
			"createdAt": "2017-10-09T22:31:30.000Z",
			"id": "6",
			"input": "{\"_EXECUTION_NAME\":\"3956fc73-d95a-43f6-9984-22f5e5f07ffd\"}",
			"output": "{\"_EXECUTION_NAME\":\"3956fc73-d95a-43f6-9984-22f5e5f07ffd\",\"done\":true}",
			"startedAt": "2017-10-09T22:31:30.000Z",
			"state": "start",
			"stateResource": {
				"lastUpdated": "2017-10-09T22:31:30.000Z",
				"name": "cil-reliability-dashboard",
				"namespace": "raf",
				"type": "ActivityARN"
			},
			"status": "succeeded",
			"stoppedAt": "2017-10-09T22:33:13.000Z"
		}
	],
	"lastUpdated": "2017-10-10T23:13:35.200Z",
	"namespace": "raf",
	"queue": "development",
	"retries": null,
	"status": "succeeded",
	"workflowDefinition": {
		"createdAt": "2017-10-02T22:20:48.497Z",
		"id": "872ea0df-0a09-4118-868d-bc2324af8078",
		"manager": "step-functions",
		"name": "cil-reliability-dashboard:master",
		"stateMachine": {
			"StartAt": "start",
			"States": {
				"start": {
					"End": true,
					"Resource": "cil-reliability-dashboard",
					"Retry": null,
					"Type": "Task"
				}
			},
			"Version": "1.0"
		},
		"version": 4
	}
}
```

### CODE REVIEW

**Comment by mohit** [comment]
can we run a couple of multiverse jobs through this before merging?

File: executor/workflow_manager_sfn.go [hunk 1]
```diff
@@ -349,107 +349,143 @@ func (wm *SFNWorkflowManager) UpdateWorkflowStatus(workflow *models.Workflow) er
...
-						stateToJob[*currentState].StatusReason = strings.Join(reasonLines, "\n")
+				job.Status = models.JobStatusFailed
+				job.StoppedAt = strfmt.DateTime(aws.TimeValue(evt.Timestamp))
+				if details := evt.ActivityFailedEventDetails; details != nil {
+					reasonLines := strings.Split(strings.TrimSpace(aws.StringValue(details.Cause)), "\n")
+					if len(reasonLines) > maxFailureReasonLines {
+						reasonLines = reasonLines[len(reasonLines)-maxFailureReasonLines:]
 					}
+					if aws.StringValue(details.Error) != "" {
+						// TODO: need more natural place to put error name...
```
Comment (mohit) [inline executor/workflow_manager_sfn.go]: Could we just put this in `job.Output`?

Comment (rgarcia) [reply executor/workflow_manager_sfn.go]: ultimately I think we should capture failure information and put it in previous attempts, e.g. this PR results in the following `reason` getting populated in the attempts array:

```
...
			"attempts": [
				{
					"reason": "sfncli.CommandTerminated",
					"startedAt": "2017-10-09T22:31:02.000Z",
					"stoppedAt": "2017-10-09T22:31:20.000Z",
					"taskARN": "ubuntu-xenial"
				}
			],
...
```

The way it gets there is through the code here that fills `StatusReason`.

The `Output` field on the `Job` feels like a strange place to put failure data. I think it'd be easier to rethink the `JobAttempt` schema a bit, e.g. switch the `ExitCode` to be a string, remove `ContainerInstanceArn`, ...

**Comment by rgarcia** [comment]
not sure how to run multiverse jobs through this, but I could canary it?

**Pull request merged**